policy-controller/k8s/index/src/pod.rs

```diff
@@ -0,0 +1,367 @@
+                    .set(ns_name, pod_entry.key(), pod_lookups)
+                    .expect("pod must not already exist");
+
+                pod_entry.insert(pod);
+
+                Ok(())
+            }
+
+            HashEntry::Occupied(mut entry) => {
+                debug_assert!(
+                    lookups.contains(&ns_name, entry.key()),
+                    "pod must exist in lookups"
+                );
+
+                // Labels can be updated at runtime (even though that's kind of weird). If the
+                // labels have changed, then we relink servers to pods in case label selections have
+                // changed.
+                let p = entry.get_mut();
+                if p.labels.as_ref() != &pod.metadata.labels {
+                    p.labels = pod.metadata.labels.into();
+                    p.link_servers(servers);
+                }
+
+                // Note that the default-allow annotation may not be changed at runtime.
+                Ok(())
+            }
+        }
+    }
+
+    /// Extracts port information from a pod spec.
+    fn extract_ports(
+        spec: k8s::PodSpec,
+        server_rx: ServerRx,
+        kubelet: KubeletIps,
+    ) -> (PodPorts, HashMap<u16, lookup::Rx>) {
+        let mut ports = PodPorts::default();
+        let mut lookups = HashMap::new();
+
+        for container in spec.containers.into_iter() {
+            for p in container.ports.into_iter() {
+                if p.protocol.map(|p| p == "TCP").unwrap_or(true) {
+                    let port = p.container_port as u16;
+                    if ports.by_port.contains_key(&port) {
+                        debug!(port, "Port duplicated");
+                        continue;
+                    }
+
+                    let (server_tx, rx) = watch::channel(server_rx.clone());
+                    let pod_port = Port {
+                        server_name: None,
+                        server_tx,
+                    };
+
+                    trace!(%port, name = ?p.name, "Adding port");
+                    if let Some(name) = p.name {
+                        ports.by_name.entry(name).or_default().push(port);
+                    }
+
+                    ports.by_port.insert(port, pod_port);
+                    lookups.insert(port, lookup::Rx::new(kubelet.clone(), rx));
+                }
+            }
+        }
+
+        (ports, lookups)
+    }
+
+    pub(crate) fn link_servers(&mut self, servers: &SrvIndex) {
+        for pod in self.index.values_mut() {
+            pod.link_servers(servers)
+        }
+    }
+
+    pub(crate) fn reset_server(&mut self, name: &str) {
+        for (pod_name, pod) in self.index.iter_mut() {
+            let rx = pod.default_allow_rx.clone();
+            for (p, port) in pod.ports.by_port.iter_mut() {
+                if port
+                    .server_name
+                    .as_ref()
+                    .map(|n| n == name)
+                    .unwrap_or(false)
+                {
```

Comment:
nit: can't we just write
```suggestion
                if port.server_name.as_ref() == Some(name) {
```
or is there a reason that doesn't work?